File: apps/web/lib/seed.ts
import { PrismaClient } from "@prisma/client";
import nacl from "tweetnacl";
import { setPrivateKey } from "./store";
import { toHex } from "./crypto";

const prisma = new PrismaClient();

export async function seed() {
  console.log("Seeding database...");

  // Clean existing data
  await prisma.trace.deleteMany();
  await prisma.booking.deleteMany();
  await prisma.healthMetric.deleteMany();
  await prisma.calendarEvent.deleteMany();
  await prisma.policy.deleteMany();
  await prisma.capability.deleteMany();
  await prisma.human.deleteMany();

  // Generate deterministic keypairs from handle (survive restarts)
  const pariKp = nacl.sign.keyPair.fromSeed(
    new TextEncoder().encode("pari".padEnd(32, "\0"))
  );
  const alexKp = nacl.sign.keyPair.fromSeed(
    new TextEncoder().encode("alex".padEnd(32, "\0"))
  );
  const drSmithKp = nacl.sign.keyPair.fromSeed(
    new TextEncoder().encode("dr_smith".padEnd(32, "\0"))
  );

  // Store private keys in runtime
  setPrivateKey("pari", pariKp.secretKey);
  setPrivateKey("alex", alexKp.secretKey);
  setPrivateKey("dr_smith", drSmithKp.secretKey);

  const baseUrl = process.env.NEXT_PUBLIC_BASE_URL || "http://localhost:3000";

  // Create humans
  const pari = await prisma.human.create({
    data: {
      handle: "pari",
      displayName: "Pari",
      publicKey: toHex(pariKp.publicKey),
      endpointUrl: `${baseUrl}/api/u/pari`,
    },
  });

  const alex = await prisma.human.create({
    data: {
      handle: "alex",
      displayName: "Alex",
      publicKey: toHex(alexKp.publicKey),
      endpointUrl: `${baseUrl}/api/u/alex`,
    },
  });

  const drSmith = await prisma.human.create({
    data: {
      handle: "dr_smith",
      displayName: "Dr. Smith",
      publicKey: toHex(drSmithKp.publicKey),
      endpointUrl: `${baseUrl}/api/u/dr_smith`,
    },
  });

  // Capabilities for alex
  await prisma.capability.createMany({
    data: [
      {
        humanId: alex.id,
        name: "schedule_propose",
        description: "Propose meeting times",
      },
      {
        humanId: alex.id,
        name: "schedule_counter",
        description: "Counter-propose meeting times",
      },
      {
        humanId: alex.id,
        name: "schedule_confirm",
        description: "Confirm a meeting booking",
      },
      {
        humanId: alex.id,
        name: "execute_trade",
        description: "Execute a financial trade",
      },
    ],
  });

  // Capabilities for pari
  await prisma.capability.createMany({
    data: [
      {
        humanId: pari.id,
        name: "schedule_propose",
        description: "Propose meeting times",
      },
      {
        humanId: pari.id,
        name: "schedule_confirm",
        description: "Confirm a meeting booking",
      },
      {
        humanId: pari.id,
        name: "health_summary",
        description: "View health analytics summary",
      },
      {
        humanId: pari.id,
        name: "health.anomaly_alert",
        description: "Receive and triage a health anomaly alert from wearable data",
      },
    ],
  });

  // Capabilities for dr_smith
  await prisma.capability.createMany({
    data: [
      {
        humanId: drSmith.id,
        name: "triage.intake_and_schedule",
        description: "Run intake questions and schedule a clinic appointment",
      },
    ],
  });

  // Policies for alex
  await prisma.policy.createMany({
    data: [
      {
        humanId: alex.id,
        capabilityName: "schedule_propose",
        allowedCallersJson: JSON.stringify(["pari"]),
        requiredScopesJson: JSON.stringify([]),
      },
      {
        humanId: alex.id,
        capabilityName: "schedule_counter",
        allowedCallersJson: JSON.stringify(["pari"]),
        requiredScopesJson: JSON.stringify([]),
      },
      {
        humanId: alex.id,
        capabilityName: "schedule_confirm",
        allowedCallersJson: JSON.stringify(["pari"]),
        requiredScopesJson: JSON.stringify(["propose_meeting"]),
      },
      {
        humanId: alex.id,
        capabilityName: "execute_trade",
        allowedCallersJson: JSON.stringify([]),
        requiredScopesJson: JSON.stringify(["trade_exec"]),
      },
    ],
  });

  // Policies for pari
  await prisma.policy.createMany({
    data: [
      {
        humanId: pari.id,
        capabilityName: "health_summary",
        allowedCallersJson: JSON.stringify(["dr_smith"]),
        requiredScopesJson: JSON.stringify([]),
        paymentRequired: true,
        priceCents: 500,
      },
      {
        humanId: pari.id,
        capabilityName: "schedule_propose",
        allowedCallersJson: JSON.stringify(["*"]),
      },
      {
        humanId: pari.id,
        capabilityName: "schedule_confirm",
        allowedCallersJson: JSON.stringify(["*"]),
        requiredScopesJson: JSON.stringify(["propose_meeting"]),
      },
      {
        humanId: pari.id,
        capabilityName: "health.anomaly_alert",
        allowedCallersJson: JSON.stringify(["pari"]),
        requiredScopesJson: JSON.stringify(["health:write"]),
      },
    ],
  });

  // Policies for dr_smith
  await prisma.policy.createMany({
    data: [
      {
        humanId: drSmith.id,
        capabilityName: "triage.intake_and_schedule",
        allowedCallersJson: JSON.stringify(["pari"]),
        requiredScopesJson: JSON.stringify(["triage:write"]),
      },
    ],
  });

  // Alex's calendar: busy slots next week
  const nextMon = getNextMonday();
  await prisma.calendarEvent.createMany({
    data: [
      {
        humanId: alex.id,
        title: "Team standup",
        startTs: setTime(nextMon, 9, 0).toISOString(),
        endTs: setTime(nextMon, 9, 30).toISOString(),
      },
      {
        humanId: alex.id,
        title: "Lunch with team",
        startTs: setTime(nextMon, 12, 0).toISOString(),
        endTs: setTime(nextMon, 13, 0).toISOString(),
      },
      {
        humanId: alex.id,
        title: "Design review",
        startTs: setTime(addDays(nextMon, 1), 14, 0).toISOString(),
        endTs: setTime(addDays(nextMon, 1), 15, 30).toISOString(),
      },
      {
        humanId: alex.id,
        title: "1:1 with manager",
        startTs: setTime(addDays(nextMon, 2), 10, 0).toISOString(),
        endTs: setTime(addDays(nextMon, 2), 10, 30).toISOString(),
      },
    ],
  });

  // Pari's health metrics for last 30 days
  const today = new Date();
  const healthData = [];
  for (let i = 29; i >= 0; i--) {
    const date = new Date(today);
    date.setDate(date.getDate() - i);
    const dayStr = date.toISOString().split("T")[0];

    // Generate realistic-ish data with some variance
    const baseSleep = 6.5 + Math.sin(i * 0.3) * 1.5;
    const baseSteps = 7000 + Math.sin(i * 0.5) * 3000;
    const adherence = Math.random() > 0.15; // 85% adherence
    const symptom = 3 + Math.sin(i * 0.7) * 2 + (i === 10 ? 4 : 0); // spike on day 10

    healthData.push({
      humanId: pari.id,
      date: dayStr,
      sleepHours: Math.round(Math.max(4, Math.min(9, baseSleep)) * 10) / 10,
      steps: Math.round(Math.max(2000, baseSteps)),
      medAdherence: adherence,
      symptomScore: Math.round(Math.max(1, Math.min(10, symptom)) * 10) / 10,
    });
  }
  await prisma.healthMetric.createMany({ data: healthData });

  console.log("Seed complete: pari, alex, dr_smith created with demo data.");
  return { pariKp, alexKp, drSmithKp };
}

/* ── Date helpers ── */

function getNextMonday(): Date {
  const d = new Date();
  const day = d.getDay();
  const diff = ((8 - day) % 7) || 7;
  d.setDate(d.getDate() + diff);
  d.setHours(0, 0, 0, 0);
  return d;
}

function addDays(d: Date, n: number): Date {
  const r = new Date(d);
  r.setDate(r.getDate() + n);
  return r;
}

function setTime(d: Date, h: number, m: number): Date {
  const r = new Date(d);
  r.setHours(h, m, 0, 0);
  return r;
}

// Run directly if called from CLI
if (require.main === module) {
  seed()
    .then(() => process.exit(0))
    .catch((e) => {
      console.error(e);
      process.exit(1);
    });
}

File: apps/api/app/seed.py
"""Database seeder (ported from apps/web/lib/seed.ts + ensureSeed.ts)."""

from __future__ import annotations

import json
import math
import random
from datetime import datetime, timedelta

from .db import get_supabase, set_private_key
from .crypto import to_hex, keypair_from_seed
from .config import get_settings

_seeded = False


async def ensure_seed() -> None:
    """Idempotent seed check. Safe to call on every request."""
    global _seeded
    if _seeded:
        return

    db = get_supabase()
    res = db.table("humans").select("id", count="exact").limit(1).execute()
    count = res.count or 0

    if count > 0:
        await _rehydrate_keys()
        _seeded = True
        return

    await seed()
    _seeded = True


async def _rehydrate_keys() -> None:
    """Regenerate deterministic keypairs from handles (demo only)."""
    db = get_supabase()
    res = db.table("humans").select("id, handle").execute()

    for h in (res.data or []):
        handle = h["handle"]
        seed_bytes = handle.encode("utf-8").ljust(32, b"\x00")[:32]
        pk, sk = keypair_from_seed(seed_bytes)
        set_private_key(handle, sk)

        # Update public key to match regenerated keypair
        db.table("humans").update({"public_key": to_hex(pk)}).eq("id", h["id"]).execute()


async def seed() -> None:
    """Full database seed with demo data."""
    db = get_supabase()

    # Clean existing data (order matters for FK constraints)
    for table in ["anomaly_alerts", "traces", "bookings", "health_metrics", "calendar_events", "policies", "capabilities", "humans"]:
        db.table(table).delete().neq("id", "00000000-0000-0000-0000-000000000000").execute()

    # Generate deterministic keypairs
    pari_pk, pari_sk = keypair_from_seed("pari".encode().ljust(32, b"\x00")[:32])
    alex_pk, alex_sk = keypair_from_seed("alex".encode().ljust(32, b"\x00")[:32])
    dr_smith_pk, dr_smith_sk = keypair_from_seed("dr_smith".encode().ljust(32, b"\x00")[:32])

    set_private_key("pari", pari_sk)
    set_private_key("alex", alex_sk)
    set_private_key("dr_smith", dr_smith_sk)

    base_url = get_settings().base_url

    # Create humans
    pari = db.table("humans").insert({
        "handle": "pari",
        "display_name": "Pari",
        "public_key": to_hex(pari_pk),
        "endpoint_url": f"{base_url}/api/u/pari",
    }).execute().data[0]

    alex = db.table("humans").insert({
        "handle": "alex",
        "display_name": "Alex",
        "public_key": to_hex(alex_pk),
        "endpoint_url": f"{base_url}/api/u/alex",
    }).execute().data[0]

    dr_smith = db.table("humans").insert({
        "handle": "dr_smith",
        "display_name": "Dr. Smith",
        "public_key": to_hex(dr_smith_pk),
        "endpoint_url": f"{base_url}/api/u/dr_smith",
    }).execute().data[0]

    # Capabilities for alex
    db.table("capabilities").insert([
        {"human_id": alex["id"], "name": "schedule_propose", "description": "Propose meeting times"},
        {"human_id": alex["id"], "name": "schedule_counter", "description": "Counter-propose meeting times"},
        {"human_id": alex["id"], "name": "schedule_confirm", "description": "Confirm a meeting booking"},
        {"human_id": alex["id"], "name": "execute_trade", "description": "Execute a financial trade"},
    ]).execute()

    # Capabilities for pari
    db.table("capabilities").insert([
        {"human_id": pari["id"], "name": "schedule_propose", "description": "Propose meeting times"},
        {"human_id": pari["id"], "name": "schedule_confirm", "description": "Confirm a meeting booking"},
        {"human_id": pari["id"], "name": "health_summary", "description": "View health analytics summary"},
        {"human_id": pari["id"], "name": "health.anomaly_alert", "description": "Receive and triage a health anomaly alert from wearable data"},
    ]).execute()

    # Capabilities for dr_smith
    db.table("capabilities").insert([
        {"human_id": dr_smith["id"], "name": "triage.intake_and_schedule", "description": "Run intake questions and schedule a clinic appointment"},
    ]).execute()

    # Policies for alex
    db.table("policies").insert([
        {"human_id": alex["id"], "capability_name": "schedule_propose", "allowed_callers_json": json.dumps(["pari"]), "required_scopes_json": json.dumps([])},
        {"human_id": alex["id"], "capability_name": "schedule_counter", "allowed_callers_json": json.dumps(["pari"]), "required_scopes_json": json.dumps([])},
        {"human_id": alex["id"], "capability_name": "schedule_confirm", "allowed_callers_json": json.dumps(["pari"]), "required_scopes_json": json.dumps(["propose_meeting"])},
        {"human_id": alex["id"], "capability_name": "execute_trade", "allowed_callers_json": json.dumps([]), "required_scopes_json": json.dumps(["trade_exec"])},
    ]).execute()

    # Policies for pari
    db.table("policies").insert([
        {"human_id": pari["id"], "capability_name": "health_summary", "allowed_callers_json": json.dumps(["dr_smith"]), "required_scopes_json": json.dumps([]), "payment_required": True, "price_cents": 500},
        {"human_id": pari["id"], "capability_name": "schedule_propose", "allowed_callers_json": json.dumps(["*"]), "required_scopes_json": json.dumps([])},
        {"human_id": pari["id"], "capability_name": "schedule_confirm", "allowed_callers_json": json.dumps(["*"]), "required_scopes_json": json.dumps(["propose_meeting"])},
        {"human_id": pari["id"], "capability_name": "health.anomaly_alert", "allowed_callers_json": json.dumps(["pari"]), "required_scopes_json": json.dumps(["health:write"])},
    ]).execute()

    # Policies for dr_smith
    db.table("policies").insert([
        {"human_id": dr_smith["id"], "capability_name": "triage.intake_and_schedule", "allowed_callers_json": json.dumps(["pari"]), "required_scopes_json": json.dumps(["triage:write"])},
    ]).execute()

    # Alex's calendar: busy slots next week
    next_mon = _get_next_monday()
    db.table("calendar_events").insert([
        {"human_id": alex["id"], "title": "Team standup", "start_ts": _set_time(next_mon, 9, 0).isoformat(), "end_ts": _set_time(next_mon, 9, 30).isoformat()},
        {"human_id": alex["id"], "title": "Lunch with team", "start_ts": _set_time(next_mon, 12, 0).isoformat(), "end_ts": _set_time(next_mon, 13, 0).isoformat()},
        {"human_id": alex["id"], "title": "Design review", "start_ts": _set_time(next_mon + timedelta(days=1), 14, 0).isoformat(), "end_ts": _set_time(next_mon + timedelta(days=1), 15, 30).isoformat()},
        {"human_id": alex["id"], "title": "1:1 with manager", "start_ts": _set_time(next_mon + timedelta(days=2), 10, 0).isoformat(), "end_ts": _set_time(next_mon + timedelta(days=2), 10, 30).isoformat()},
    ]).execute()

    # Pari's health metrics for last 30 days
    today = datetime.now()
    health_data = []
    for i in range(29, -1, -1):
        date = today - timedelta(days=i)
        day_str = date.strftime("%Y-%m-%d")

        base_sleep = 6.5 + math.sin(i * 0.3) * 1.5
        base_steps = 7000 + math.sin(i * 0.5) * 3000
        adherence = random.random() > 0.15
        symptom = 3 + math.sin(i * 0.7) * 2 + (4 if i == 10 else 0)

        health_data.append({
            "human_id": pari["id"],
            "date": day_str,
            "sleep_hours": round(max(4, min(9, base_sleep)), 1),
            "steps": round(max(2000, base_steps)),
            "med_adherence": adherence,
            "symptom_score": round(max(1, min(10, symptom)), 1),
        })

    db.table("health_metrics").insert(health_data).execute()

    print("Seed complete: pari, alex, dr_smith created with demo data.")


def _get_next_monday() -> datetime:
    d = datetime.now()
    days_ahead = (7 - d.weekday()) % 7
    if days_ahead == 0:
        days_ahead = 7
    d = d + timedelta(days=days_ahead)
    return d.replace(hour=0, minute=0, second=0, microsecond=0)


def _set_time(d: datetime, h: int, m: int) -> datetime:
    return d.replace(hour=h, minute=m, second=0, microsecond=0)
